swap/swap.go

```diff
@@ -1,98 +1,451 @@
+		address:    crypto.PubkeyToAddress(*pubkey),
+	}
+}
+
+// DeploySuccess is for convenience log output
+func (s *Swap) DeploySuccess() string {
+	return fmt.Sprintf("contract: %s, owner: %s, deposit: %v, signer: %x", s.owner.Contract.Hex(), s.owner.address.Hex(), s.params.InitialDepositAmount, s.owner.publicKey)
 }
 
-//Swap implements the protocols.Balance interface
-//Add is the (sole) accounting function
+// Add is the (sole) accounting function
+// Swap implements the protocols.Balance interface
 func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
-	//load existing balances from the state store
-	err = s.loadState(peer)
+	// load existing balances from the state store
+	err = s.loadBalance(peer.ID())
 	if err != nil && err != state.ErrNotFound {
-		return
+		return fmt.Errorf("error while loading balance for peer %s", peer.ID().String())
+	}
+
+	// Check if balance with peer is over the disconnect threshold
+	// It is the creditor who triggers the disconnect from a overdraft creditor,
+	// thus we check for a positive value
+	if s.balances[peer.ID()] >= s.disconnectThreshold {
+		// if so, return error in order to abort the transfer
+		return fmt.Errorf("balance for peer %s is over the disconnect threshold %d, disconnecting", peer.ID().String(), s.disconnectThreshold)
+	}
+
+	// calculate new balance
+	var newBalance int64
+	newBalance, err = s.updateBalance(peer.ID(), amount)
+	if err != nil {
+		return err
 	}
+
+	// Check if balance with peer crosses the threshold
+	// It is the peer with a negative balance who sends a cheque, thus we check
+	// that the balance is *below* the threshold
+	if newBalance <= -s.paymentThreshold {
+		//if so, send cheque
+		log.Warn("balance for peer went over the payment threshold, sending cheque", "peer", peer.ID().String(), "payment threshold", s.paymentThreshold)
+		return s.sendCheque(peer.ID())
+	}
+
+	return nil
+}
+
+func (s *Swap) updateBalance(peer enode.ID, amount int64) (int64, error) {
 	//adjust the balance
 	//if amount is negative, it will decrease, otherwise increase
-	s.balances[peer.ID()] += amount
+	s.balances[peer] += amount
 	//save the new balance to the state store
-	peerBalance := s.balances[peer.ID()]
-	err = s.stateStore.Put(peer.ID().String(), &peerBalance)
+	peerBalance := s.balances[peer]
+	err := s.store.Put(balanceKey(peer), &peerBalance)
+	if err != nil {
+		return 0, fmt.Errorf("error while storing balance for peer %s", peer.String())
+	}
+	log.Debug("balance for peer after accounting", "peer", peer.String(), "balance", strconv.FormatInt(peerBalance, 10))
+	return peerBalance, err
+}
 
-	log.Debug(fmt.Sprintf("balance for peer %s: %s", peer.ID().String(), strconv.FormatInt(peerBalance, 10)))
-	return err
+// loadBalance loads balances from the state store (persisted)
+func (s *Swap) loadBalance(peer enode.ID) (err error) {
+	var peerBalance int64
+	//only load if the current instance doesn't already have this peer's
```

Comment:
remove comment. code is self explanatory

Reply:
fixed